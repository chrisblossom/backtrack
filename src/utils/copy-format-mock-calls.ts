/**
 * copy.mock.calls is async so order is not guaranteed for expected results
 * Converts to plain object
 *
 * For testing purposes only
 *
 * meant to be used as a snapshot with jest-serializer-path
 */

type CopyCalls = [string, string, {}];

function copyFormatMockCalls(calls: ReadonlyArray<CopyCalls>) {
	const result = calls.reduce((acc, call) => {
		const src = call[0];
		const dest = call[1];
		const options = call[2];

		return {
			...acc,
			[src]: {
				dest,
				options,
			},
		};
	}, {});

	return result;
}

export { copyFormatMockCalls };
