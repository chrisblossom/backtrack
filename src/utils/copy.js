/* @flow */

import fse from 'fs-extra';
import { toArray } from './object-utils';

import type { CopyFile } from '../types.js';

/**
 * See https://github.com/jprichardson/node-fs-extra/blob/master/docs/copy.md#copysrc-dest-options-callback
 * for options
 */
async function copy(files: $ReadOnlyArray<CopyFile> | CopyFile, options: {}) {
    const normalized = toArray(files);

    /**
     * Does not return any information
     */
    await Promise.all(
        normalized.map((file) => {
            return fse.copy(file.src, file.dest, options);
        }),
    );
}

export { copy };
