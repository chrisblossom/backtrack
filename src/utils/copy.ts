import path from 'path';
import fs from 'fs';
import fse from 'fs-extra';
import { readDirDeep } from 'read-dir-deep';
import { getFileHash } from './get-file-hash';
import { toArray } from './object-utils';

function stat(pathname: string) {
	return new Promise((resolve, reject) => {
		fs.stat(pathname, (error, stats) => {
			if (error) {
				reject(error);

				return;
			}

			resolve(stats);
		});
	});
}

export type File = Readonly<{
	src: string;
	dest: string;

	hash?: boolean;

	/**
	 * fs-extra copy options
	 * https://github.com/jprichardson/node-fs-extra/blob/master/docs/copy.md#copysrc-dest-options-callback
	 */
	overwrite?: boolean;
	errorOnExist?: boolean;
	preserveTimestamps?: boolean;
}>;

function getHashedName(source: string, dest: string) {
	const parsed = path.parse(dest);
	const fileHash = getFileHash(source).substring(0, 8);

	const rename = `${parsed.name}.${fileHash}${parsed.ext}`;

	const result = path.join(parsed.dir, rename);

	return result;
}

/**
 * See https://github.com/jprichardson/node-fs-extra/blob/master/docs/copy.md#copysrc-dest-options-callback
 * for options
 */
function copy(files: ReadonlyArray<File> | File): Promise<void> {
	const normalized = toArray(files);

	const copyFilesResult = normalized.map(async (file) => {
		const {
			src,
			hash = false,

			overwrite = true,
			errorOnExist = true,
			preserveTimestamps = true,
		} = file;

		const fseCopyOptions = {
			overwrite,
			errorOnExist,
			preserveTimestamps,
		};

		let dest = file.dest;

		// @ts-ignore
		const isDirectory: Boolean = (await stat(src)).isDirectory();
		if (isDirectory) {
			const deepFileList: string[] = await readDirDeep(src);

			type CopyFilesToDestAcc = {
				src: string;
				dest: string;
				overwrite: boolean;
				errorOnExist: boolean;
				preserveTimestamps: boolean;
			};

			const copyFilesToDest = deepFileList.reduce(
				(acc: CopyFilesToDestAcc[], sourceFile) => {
					const absoluteSourcePath = path.resolve(src, sourceFile);

					let absoluteDestinationPath = path.resolve(
						dest,
						sourceFile,
					);

					if (hash === true) {
						absoluteDestinationPath = getHashedName(
							absoluteSourcePath,
							absoluteDestinationPath,
						);
					}

					return [
						...acc,
						{
							src: absoluteSourcePath,
							dest: absoluteDestinationPath,
							...fseCopyOptions,
						},
					];
				},
				[],
			);

			/**
			 * copy all directory files
			 */
			return copy(copyFilesToDest);
		}

		if (hash === true) {
			dest = getHashedName(src, file.dest);
		}

		return fse.copy(src, dest, fseCopyOptions);
	});

	return Promise.all(copyFilesResult).then(() => {
		/**
		 * fse.copy does not return any values. explicitly return undefined for promise.all
		 */
		return undefined;
	});
}

export { copy };
