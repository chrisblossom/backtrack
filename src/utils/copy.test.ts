import path from 'path';
import { copyFormatMockCalls } from './copy-format-mock-calls';
import { File } from './copy';

const copy = (files: File[] | File) => require('./copy').copy(files);

describe('copy', () => {
	let fseCopySpy: any;
	const cwd = process.cwd();
	const workingCwd = path.resolve(__dirname, '__sandbox__');

	beforeEach(() => {
		const fse = require('fs-extra');
		fseCopySpy = jest
			.spyOn(fse, 'copy')
			.mockImplementation(() => Promise.resolve());

		process.chdir(workingCwd);
	});

	afterEach(() => {
		process.chdir(cwd);
		jest.restoreAllMocks();
	});

	test('copies files', async () => {
		const files = {
			src: path.resolve('files1/test-file.js'),
			dest: path.resolve('build/static/test-file.js'),
		};

		const result = await copy(files);

		expect(result).toEqual(undefined);
		expect(copyFormatMockCalls(fseCopySpy.mock.calls)).toMatchSnapshot();
	});

	test('passes options to fse.copy', async () => {
		const files = {
			src: path.normalize('./files1/test-file.js'),
			dest: path.normalize('./build/static/test-file.js'),

			overwrite: false,
			errorOnExist: false,
			preserveTimestamps: false,
		};

		await copy(files);

		expect(copyFormatMockCalls(fseCopySpy.mock.calls)).toMatchSnapshot();
	});

	test('adds hash to dest with hash: true', async () => {
		const files = {
			src: path.normalize('./files1/test-file.js'),
			dest: path.normalize('./build/static/test-file.js'),
			hash: true,
		};

		await copy(files);

		expect(copyFormatMockCalls(fseCopySpy.mock.calls)).toMatchSnapshot();
	});

	test('copies directory of files', async () => {
		const files = {
			src: path.normalize('./files1'),
			dest: path.normalize('./build/static'),
			overwrite: false,
			errorOnExist: false,
			preserveTimestamps: false,
		};

		await copy(files);

		expect(copyFormatMockCalls(fseCopySpy.mock.calls)).toMatchSnapshot();
	});

	test('copies directory of files with hashes', async () => {
		const files = {
			src: path.normalize('./files1'),
			dest: path.normalize('./build/static'),
			hash: true,
		};

		await copy(files);

		expect(copyFormatMockCalls(fseCopySpy.mock.calls)).toMatchSnapshot();
	});
});
