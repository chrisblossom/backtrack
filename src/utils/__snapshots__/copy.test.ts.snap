// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copy adds hash to dest with hash: true 1`] = `
{
  "files1/test-file.js": {
    "dest": "build/static/test-file.b2c22cd6.js",
    "options": {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  },
}
`;

exports[`copy copies directory of files 1`] = `
{
  "<PROJECT_ROOT>/files1/nested/nested-inside/inside.js": {
    "dest": "<PROJECT_ROOT>/build/static/nested/nested-inside/inside.js",
    "options": {
      "errorOnExist": false,
      "overwrite": false,
      "preserveTimestamps": false,
    },
  },
  "<PROJECT_ROOT>/files1/nested/other.js": {
    "dest": "<PROJECT_ROOT>/build/static/nested/other.js",
    "options": {
      "errorOnExist": false,
      "overwrite": false,
      "preserveTimestamps": false,
    },
  },
  "<PROJECT_ROOT>/files1/test-file.js": {
    "dest": "<PROJECT_ROOT>/build/static/test-file.js",
    "options": {
      "errorOnExist": false,
      "overwrite": false,
      "preserveTimestamps": false,
    },
  },
}
`;

exports[`copy copies directory of files with hashes 1`] = `
{
  "<PROJECT_ROOT>/files1/nested/nested-inside/inside.js": {
    "dest": "<PROJECT_ROOT>/build/static/nested/nested-inside/inside.f7d8aade.js",
    "options": {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  },
  "<PROJECT_ROOT>/files1/nested/other.js": {
    "dest": "<PROJECT_ROOT>/build/static/nested/other.4a98043a.js",
    "options": {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  },
  "<PROJECT_ROOT>/files1/test-file.js": {
    "dest": "<PROJECT_ROOT>/build/static/test-file.b2c22cd6.js",
    "options": {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  },
}
`;

exports[`copy copies files 1`] = `
{
  "<PROJECT_ROOT>/files1/test-file.js": {
    "dest": "<PROJECT_ROOT>/build/static/test-file.js",
    "options": {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  },
}
`;

exports[`copy passes options to fse.copy 1`] = `
{
  "files1/test-file.js": {
    "dest": "build/static/test-file.js",
    "options": {
      "errorOnExist": false,
      "overwrite": false,
      "preserveTimestamps": false,
    },
  },
}
`;
